crates/oxc_minifier/src/peephole/minimize_statements.rs

```diff
@@ -443,6 +446,254 @@ impl<'a> PeepholeOptimizations {
+            let init_val = match init_iter.next() {
+                None | Some(ArrayExpressionElement::Elision(_)) => {
+                    ArrayExpressionElement::from(ctx.ast.void_0(SPAN))
+                }
+                Some(init_val) => init_val,
+            };
+
+            match id_item {
+                // `[a = b] = [??]`
+                Some(BindingPattern::AssignmentPattern(mut pattern)) => {
+                    let init_value_type = init_val.value_type(ctx);
+                    if init_value_type.is_undefined() {
+                        // if value is undefined, `[a = b] = [undefined]` => `a = b`
+                        result.push(ctx.ast.variable_declarator(
+                            pattern.span(),
+                            decl.kind,
+                            pattern.left.take_in(ctx.ast),
+                            NONE,
+                            Some(pattern.right.take_in(ctx.ast)),
+                            decl.definite,
+                        ));
+                    } else if init_value_type.is_undetermined() {
+                        // `[a = b] = [c]` where c is undetermined => `[a = b] = [c]`
+                        result.push(ctx.ast.variable_declarator(
+                            pattern.span(),
+                            decl.kind,
+                            ctx.ast.binding_pattern_array_pattern(
+                                decl.span,
+                                ctx.ast.vec1(Some(BindingPattern::AssignmentPattern(pattern))),
+                                NONE,
+                            ),
+                            NONE,
+                            Some(ctx.ast.expression_array(init_val.span(), ctx.ast.vec1(init_val))),
+                            decl.definite,
+                        ));
+                    } else {
+                        // if value is determined, `[a = b] = [c]` => `a = c`
+                        result.push(ctx.ast.variable_declarator(
+                            pattern.span(),
+                            decl.kind,
+                            pattern.left.take_in(ctx.ast),
+                            NONE,
+                            Some(init_val.into_expression()),
+                            decl.definite,
+                        ));
+                    }
+                }
+                // `[a, b] = [c, d]` => `a = c, b = d`
+                Some(id) => {
+                    result.push(ctx.ast.variable_declarator(
+                        id.span(),
+                        decl.kind,
+                        id,
+                        NONE,
+                        Some(init_val.into_expression()),
+                        decl.definite,
+                    ));
+                }
+                // `[,] = [??]` => `void [] = [??]`
+                None => {
+                    // `[,] = [,]` => `void [] = [,]`
+                    if !init_val.value_type(ctx).is_undefined() {
```

Comment:
We need to keep the RHS when `init_val` has sideeffects here as well.

Reply:
like always thx, i didn't think about that, i'm going to update it